pkg/api/steve/collections/collections.go

```diff
@@ -0,0 +1,699 @@
+// Package collections supports the /apis/collections.cattle.io/v1alpha1 endpoint for listing resources by project.
+package collections
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/gorilla/mux"
+	"github.com/rancher/rancher/pkg/controllers/managementagent/nslabels"
+	"github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/rancher/pkg/wrangler"
+	"github.com/rancher/wrangler/pkg/apply"
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
+	authzv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	clientauthzv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	apiregv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+)
+
+const (
+	// Group is the APIGroup for thie APIService.
+	Group = "collections.cattle.io"
+	// ParentLabel is the label assigned to namespaces that represent Projects.
+	ParentLabel = "cattle.io/parent"
+
+	apiServiceName            = version + "." + Group
+	groupVersion              = Group + "/" + version
+	version                   = "v1alpha1"
+	workers                   = int64(3)
+	queryKey                  = "fieldSelector"
+	projectsOrNamespacesKey   = "prjorns"
```

Comment:
I think this was meant for saving characters in the URL but in hindsight it made the implementation very complex, will look into having separate parameters.

Reply:
I played around with this and added the changes as a separate commit so it would be easy to see the difference. It actually adds some complexity in some annoying ways, 1) since this isn't the primary query key (`fieldSelector` is) then it needs a distinct separator that isn't `&` or `,` to separate the two keys, and 2) now there are two possible operators, equality/nonequality for projects and the same for namespaces. I updated the tests to make them pass but I would need to add more test cases to cover that second caveat better. Let me know what you think of this direction.

Reply:
To support something like `fieldSelector=projects=p1,p2+namespaces=n1,n2` which I attempted in f820be06f46e24616ef117f87ed6271de4d4d7f9 we have to decide what to do when selectors overlap and the operators are in conflict with each other. For example, if namespace n1 is part of project p1 and I have `fieldSelector=projects=p1+namespaces!=n1`, what's the result? Do projects override namespaces or vice versa, does != override = or vice versa? This isn't something the UI currently supports or has asked for. I'm leaning toward keeping things simple-ish by using just a single parameter for both.

Reply:
Given the complexity I would tend to agree with you. I think we get less value from this because projects are also namespaces now in the downstream - so there's no chance for overlap. 

I'm ok with using one parameter, but I'd like to request that:
- We make it clear in the docs/project-readme that filtering by `projNS=p-1234xz` will filter on the project if it's a project, and that you can't use the parameter to filter on the resources in the downstream namespace by the same name (i.e. you can't use this endpoint to get the resources in the downstream project namespace)
- We come up with something more clear than `pjorns` as the name for the parameter. But we can do this whenever since it's a simple name change. 

LMK if you disagree.

Reply:
Sounds good to me.

For naming, I'll go with `projectsornamespaces` unless anyone proposes something better.